request.go

```diff
@@ -27,3 +33,67 @@ func (r *request) encode(pe packetEncoder) (err error) {
 	}
 	return pe.pop()
 }
+
+func (r *request) decode(pd packetDecoder) (err error) {
+	var key int16
+	if key, err = pd.getInt16(); err != nil {
+		return err
+	}
+	var version int16
+	if version, err = pd.getInt16(); err != nil {
+		return err
+	}
+	if r.correlationID, err = pd.getInt32(); err != nil {
+		return err
+	}
+	r.clientID, err = pd.getString()
+
+	r.body = allocateBody(key, version)
+	if r.body == nil {
+		return fmt.Errorf("Unknown request key: %d", key)
+	}
+	return r.body.decode(pd)
+}
+
+func decodeRequest(r io.Reader) (req *request, err error) {
+	lengthBytes := make([]byte, 4)
+	if _, err := io.ReadFull(r, lengthBytes); err != nil {
+		return nil, err
+	}
+
+	var length int32
+	if length, err = (&realDecoder{raw: lengthBytes}).getInt32(); err != nil {
+		return nil, err
+	}
+
+	encodedReq := make([]byte, length)
```

Comment:
you'll want to add a validation the length is reasonable, something like https://github.com/Shopify/sarama/blob/master/response_header.go#L15-L17